File: server/Sources/MoqCompanionAI/Providers/AIProvider.swift
/// Protocol for AI provider backends (Ollama, OpenAI, Anthropic, etc.).
public protocol AIProvider: Sendable {
    /// Unique provider identifier (e.g. "ollama", "openai", "anthropic").
    var id: String { get }

    /// Human-readable display name.
    var displayName: String { get }

    /// Whether this provider runs locally or is cloud-hosted.
    var kind: ProviderKind { get }

    /// AI capabilities this provider supports.
    var capabilities: [AICapability] { get }

    /// Check if the provider is currently reachable and configured.
    func checkAvailability() async -> Bool

    /// Validate that the provider's configuration is correct.
    func validateConfig() async -> ValidateConfigResponse

    /// Send a structured prompt and return a raw text completion.
    func complete(prompt: String, model: String?, temperature: Double?) async throws -> AICompletionResult
}

/// Result from a provider completion call.
public struct AICompletionResult: Sendable {
    public let text: String
    public let model: String
    public let usage: UsageInfo?

    public init(text: String, model: String, usage: UsageInfo? = nil) {
        self.text = text
        self.model = model
        self.usage = usage
    }
}

File: studio/studio-ai/src/main/kotlin/com/moqserver/studio/ai/AIProvider.kt
package com.moqserver.studio.ai

/** A single AI backend that can complete text prompts. */
interface AIProvider {
    val id: String
    val displayName: String
    val kind: AIProviderKind
    val capabilities: Set<AIProviderCapability>

    /** Returns true if the provider is reachable and configured right now. */
    suspend fun checkAvailability(): Boolean

    /**
     * Validates configuration and returns a list of human-readable issues.
     * An empty list means the configuration is valid.
     */
    suspend fun validateConfig(): List<String>

    /** Sends a text prompt and returns the raw completion result. */
    suspend fun complete(
        prompt: String,
        model: String? = null,
        temperature: Double? = null,
    ): AICompletionResult
}

enum class AIProviderKind { LOCAL, HOSTED }

enum class AIProviderCapability {
    ANALYZE_SPEC,
    GENERATE_VARIANTS,
    REFINE_PROJECT,
}

data class AICompletionResult(
    val text: String,
    val model: String,
    val promptTokens: Int? = null,
    val completionTokens: Int? = null,
    val totalTokens: Int? = null,
    val latencyMs: Long? = null,
)
